### ISSUE TITLE
Download any file with Python requests library

### ISSUE THREAD

**Issue by openedhardware** [docker-slim/docker-slim#179]
I have a python code that reads an environment variable and downloads a file from the cloud storage:
```python
import os
import requests

model_name = os.environ.get('MODEL')

if model_name:
    url = "https://viso-public-storage.s3.eu-central-1.amazonaws.com/ObjDet/CPU/{}.tar.xz".format(model_name)
    print(f"Downloading - {url}")
    with open(f'/data/{model_name}.tar.gz', 'wb') as f:
        r = requests.get(url)
        f.write(r.content)
print("Finished!")
```
If `MODEL` environment variable exists, it downloads from our S3 bucket.

But once slimified, it doesn't work saying that the device or resource is busy:
```shell
ubuntu@ubuntu:~/test_docker_slim$ docker run -v /data:/data --rm --network host -e MODEL='ssd_mobilenet_v1_coco_2018_01_28' ts.slim
Downloading - https://viso-public-storage.s3.eu-central-1.amazonaws.com/ObjDet/CPU/ssd_mobilenet_v1_coco_2018_01_28.tar.xz
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 170, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/connection.py", line 73, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "/usr/local/lib/python3.7/socket.py", line 752, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
OSError: [Errno 16] Device or resource busy

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 706, in urlopen
    chunked=chunked,
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 382, in _make_request
    self._validate_conn(conn)
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 1010, in _validate_conn
    conn.connect()
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 353, in connect
    conn = self._new_conn()
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 182, in _new_conn
    self, "Failed to establish a new connection: %s" % e
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPSConnection object at 0x7fce707c1210>: Failed to establish a new connection: [Errno 16] Device or resource busy

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 756, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/retry.py", line 573, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='viso-public-storage.s3.eu-central-1.amazonaws.com', port=443): Max retries exceeded with url: /ObjDet/CPU/ssd_mobilenet_v1_coco_2018_01_28.tar.xz (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fce707c1210>: Failed to establish a new connection: [Errno 16] Device or resource busy'))

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "main.py", line 10, in <module>
    r = requests.get(url)
  File "/usr/local/lib/python3.7/site-packages/requests/api.py", line 76, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='viso-public-storage.s3.eu-central-1.amazonaws.com', port=443): Max retries exceeded with url: /ObjDet/CPU/ssd_mobilenet_v1_coco_2018_01_28.tar.xz (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fce707c1210>: Failed to establish a new connection: [Errno 16] Device or resource busy'))
```

Any idea why docker-slim doesn't suport random file download to the mounted volume?

You can easily test on your end by cloning this repo - https://github.com/openedhardware/test_docker_slim

Cheers!

**Comment by kcq** [comment]
Thanks for the repro project and the repro instructions @openedhardware ! I think this failure is due to a missing env var when you run `docker-slim`. Adding `--env MODEL='ssd_mobilenet_v1_coco_2018_01_28'` to the `build` command did the trick for me. Without that env var the `url` value in the app ends up being invalid, so the `requests.get(url)` call in the app fails when `docker-slim` executes the application in its temporary container.

Here's what the `build` command that worked for me: `docker-slim build --http-probe=false --env MODEL='ssd_mobilenet_v1_coco_2018_01_28' --mount $(pwd)/data:/data --include-path /usr/local/lib/python3.7/site-packages/certifi --show-clogs ts`

**Comment by openedhardware** [comment]
@kcq 
Yeah, it worked with a constant variable when `docker-slim build`.
But that environment variable will be changed depending on the condition.

How can we support various variables?

Thanks for your effort!

Cheers.

**Comment by kcq** [comment]
There a few options... The value of MODEL needs to reference something you have. For example, it can be a test model file name that's always there, so setting it to anything as long the generated URL points something on S3 should do the trick. Option two is to use an environment variable to set the MODEL env var for `docker-slim` (e.g., `docker-slim build --http-probe=false --env MODEL=$YOUR_CURRENT_MODEL --mount $(pwd)/data:/data --include-path /usr/local/lib/python3.7/site-packages/certifi --show-clogs ts`) where `$YOUR_CURRENT_MODEL` is set according to your conditions. Option 3 is to have the whole url in the app settable through the env vars (a few different ways to do it... including using two env vars, one for the format string and another for the model name).

**Comment by a-takahashi223** [comment]
I had a similar situation.
In my case, I give hostname as environment variable and if it is not given, it will be `localhost`.
When I ran `docker-slim build`, I didn't give the environment variable, so the app accessed `localhost`, which should have resulted in `connection refused`.
In this case, would it work if I had set the environmental variable to hostname that would return an HTTP response?
